core/src/muxing.rs

```diff
@@ -79,9 +79,6 @@ mod singleton;
 ///   `destroy_outbound` methods allow controlling these entries.
 ///
 pub trait StreamMuxer {
-    /// Type of the object that represents the raw substream where data can be read and written.
-    type Substream;
```

Comment:
What is the reasoning for using dynamic dispatching (`AsyncReadWriteBox`) instead of using an associated type that implements `AsyncRead` and `AsyncWrite`?

Reply:
Part of the motivation for this change was the idea that protocols (like `libp2p-ping`) should depend on a type that is independent from the specific muxer in question and instead are implemented on top of an opaque stream type. This would allow standalone usage of the protocol (state machines) which hopefully brings us closer to using some of the protocol logic independently from `libp2p-swarm`.

Additionally, I am not even sure an associated type would work unless we parametrize all protocols over `M: StreamMuxer` to be able to reference the concrete substream type in methods like `inject_fully_negotiated_inbound`.

Reply:
My current thinking is that it would be nice to have `libp2p-swarm` be an optional feature of protocol crates.

Reply:
> My current thinking is that it would be nice to have `libp2p-swarm` be an optional feature of protocol crates.

Protocol crates like e.g. `libp2p-gossipsub`? Would they still be using `NetworkBehaviour`?

> Additionally, I am not even sure an associated type would work unless we parametrize all protocols over `M: StreamMuxer` to be able to reference the concrete substream type in methods like `inject_fully_negotiated_inbound`?

We might decide to do dynamic dispatching at the `libp2p-swarm` level, though I am not sure we need it at the `libp2p-core` level.

Reply:
> > My current thinking is that it would be nice to have `libp2p-swarm` be an optional feature of protocol crates.
> 
> Protocol crates like e.g. `libp2p-gossipsub`? Would they still be using `NetworkBehaviour`?

They would, optionally, by dispatching to standalone state machines that only depend on `libp2p-core` and could thus be used without the `libp2p-swarm` abstractions. Without derailing the conversation too much: I see one of two avenues for making `rust-libp2p` easier for "arbitrary stream usage".

1. Re-designing the `NetworkBehaviour` trait to provide more flexibility
2. Making it easier to not use `libp2p-swarm` but still be able to re-use the existing work and protocols.

> > Additionally, I am not even sure an associated type would work unless we parametrize all protocols over `M: StreamMuxer` to be able to reference the concrete substream type in methods like `inject_fully_negotiated_inbound`?
> 
> We might decide to do dynamic dispatching at the `libp2p-swarm` level, though I am not sure we need it at the `libp2p-core` level.

That is an interesting idea, I will go and explore that.

Reply:
What do yo think of https://github.com/libp2p/rust-libp2p/pull/2648/commits/61f07e0271e143992a00463f8d0ffe50b92bd32f?

Reply:
:+1: for reintroducing `type Substream`.